lib/db_connection/connection_pool.ex

```diff
@@ -68,25 +78,36 @@ defmodule DBConnection.ConnectionPool do
       idle: nil
     }
 
+    metrics = :counters.new(2, [])
+    # when the pool starts up, it checks out and should decrement this to 0
+    :counters.put(metrics, @active_counter_idx, pool_size)
+
     codel = start_idle(now_in_native, start_poll(now_in_ms, now_in_ms, codel))
-    {:ok, {:busy, queue, codel, ts}}
+    {:ok, {:busy, queue, codel, ts, metrics}}
   end
 
   @impl GenServer
-  def handle_call({:disconnect_all, interval}, _from, {type, queue, codel, _ts}) do
+  def handle_call({:disconnect_all, interval}, _from, {type, queue, codel, _ts, metrics}) do
     ts = {System.monotonic_time(), interval}
-    {:reply, :ok, {type, queue, codel, ts}}
+    {:reply, :ok, {type, queue, codel, ts, metrics}}
+  end
+
+  def handle_call(:get_metrics, _from, {_, _, _, _, metrics} = state) do
+    active = :counters.get(metrics, @active_counter_idx)
+    waiting = :counters.get(metrics, @waiting_counter_idx)
+    {:reply, %{active: active, waiting: waiting}, state}
```

Comment:
You can return the counter to the client now, and the client reads it. :)

```suggestion
    {:reply, metrics, state}
```

Reply:
I considered that, but two big issues came up:

1. This makes the call not `:rpc` safe, since the counters returns an atomic reference that's local, and it's quite likely that you'd attempt to call this on a different node, in the situation where you have a production application on one node and a manager on another node that might be polling for metrics. (I have some testing code in this comment below for this. 
2. Passing back the counter without unwrapping imo is a bad user experience. It's not super clear which value in the counter references which. 

----------
Testing counters:
```
# node 1 iex --name node1@127.0.0.1         
defmodule CounterServer do
  use GenServer

  def start_link do
    GenServer.start_link(__MODULE__, :ok, name: __MODULE__)
  end

  def get_counter do
    GenServer.call(__MODULE__, :get)
  end


  def init(:ok) do
    # Initialize an Erlang counter
    counter = :counters.new(1, [])
    {:ok, counter}
  end

  def handle_call(:get, _from, counter) do
    # Fetch the value from the counter
    {:reply, counter, counter}
  end
end


{:ok, pid} = CounterServer.start_link(0)
```

```
# node 2 iex --name node2@127.0.0.1         

Node.connect(:"node1@127.0.0.1")

# You can define a lightweight module to delegate to the GenServer on node1
defmodule CounterClient do
  def get_counter_on_node1 do
    :rpc.call(:"node1@127.0.0.1", CounterServer, :get_counter, [])
  end
end

counter = CounterClient.get_counter_on_node1()

 :counters.get(counter, 0)
** (ArgumentError) errors were found at the given arguments:

  * 1st argument: invalid atomics reference
  * 2nd argument: out of range

    :counters.get({:atomics, #Reference<12104.2198675303.198049794.112940>}, 0)
    iex:5: (file)
```

Reply:
I think instead of returning the counter directly from one node to the other you could instead create a function that:

1. Knows what the different values mean
2. Can read the counter and pass the values nicely in a map or something like that to the other node

The alternative of letting something continuously poll the connection pool for metrics is not ideal imo.

Reply:
Those are good points. Reading the counters are quick anyway, so we can keep this as is. :)